Final version of shelley-ready cardano-crypto.js

refi93 opened the pull request:
Motivation: polish cardano-crypto.js work-in-progress version that is shelley ready into a releasable API without redundant calls/function parameters.

Changes vs the WIP version:
* removed the reduntant addressType parameter from `pack<Address Type>Address` calls
* polished bech32 api to be consitent with base58 api
* packRewardsAccountAddress renamed to packRewardAddress
* remove the option to pass the spending/staking public key into the pack<AddressType>Address call, the calls will expect the hashes directly and if the caller wants has the public key instead, they can use the `getPubKeyBlake2b224Hash` intermediate call to transform the keys to hashes which is exposed as well, the rationale being that we want an unambiguous yet flexible API for that
* added isValidShelleyAddress call to check whether a string can be considered a valid shelley address
* broke down getAddressInfo call into getAddressType and getShelleyAddressNetworkId calls as the original call did not really work for bootstrap addresses
* renamed isValidAddress to isValidBootstrapAddress as it concerns only bootstrap addresses
* grouped related functions into files which reside in the newly created `features` folder

Testing: `npm run test`

Note: this version of cardano-crypto.js has been released in npm: https://www.npmjs.com/package/cardano-crypto.js/v/5.3.6-rc.4 so we can refactor adalite to work with this library

PeterBenc commented on features/address.js:
```diff
@@ -14,10 +14,10 @@ const AddressTypes = {
   'ENTERPRISE': 0b0110,
   'BOOTSTRAP': 0b1000,
-  'REWARDS': 0b1110
+  'REWARD': 0b1110
```
I renamed this so its consistent with the `packRewadAddress` function, also in other places.

PeterBenc commented on features/address.js:
```diff
@@ -122,7 +122,7 @@ function packRewardAddress(stakingKeyHash, networkId) {
   return Buffer.concat([
-    getAddressHeader(addressType, networkId),
+    getAddressHeader(AddressTypes.REWARD, networkId),
     stakingKeyHash
```
Fixed this bug.

PeterBenc commented:
I am thinking of adding two things to this release. 

1.  export of `BootstrapAddressTypes` along with a function that takes an address string, validates that its a valid bootstrap address, and then returns the type ( by the type I mean `v1` or `v2`) 

2. function that takes `path` as a parameter, validates it and returns whether the path is Shelley or Byron

Especially the second one is very easy to implement in Adalite as well so I'm totally fine with not adding it to this release if it doesn't make sense to you. I don't think its that necessary either so pls let me know what do you think.

refi93 commented:
@PeterBenc 

1. yes, that can be added to unpackBootstrapAddress function, though I'm curious - how does it help Adalite? If it's not really needed I'm a bit skeptical because technically v1/v2 is the codename for the derivation scheme which technically is only loosely coupled to the address format so returning it may be misleading, giving the false impression that it's the same thing.
2. I would put that rather in Adalite's logic (and maybe consider if such check is actually needed and whether that isn't just a workaround for missing context from other layers of the app, pointing at the need to refactor it - we can take a look at it offline)

PeterBenc commented:
@refi93 
1. Yes adding it to `unpackBootstrapAddress` makes a lot more sense. Current use-case for it would be differentiating the address types when calculating the sizes of witnesses. Currently, I have two options, to build the witnesses, or to estimate their size based on the address type, but you make a good point that derivation schemes are only loosely coupled to the address formats. The use case for Adalite is not that important and I can surely work around it, but I was thinking that having a function for differentiating Daedalus and Icarus addresses for a Cardano library could be useful. 
I think the best option would be to implement this in Adalite, I ll create a helper function, unpack the address and return the type base on the addressAttributes.

2. Adalite is using this when creating witnesses for the ledger. Ledger returns path and signature and we iterate over these (path, sign) pairs creating Shelley or byron type witness based on the type of the given path. But yes, this functionality is really simple so I ll implement it in Adalite.

Thanks for the reply and opinion, in the end, I don't really see a strong reason to implement any of these features.

Btw. I ll test the lib with Adalite in upcoming days and approve the PR in case everything works as expected.

refi93 commented on features/address.js:
```diff
@@ -0,0 +1,246 @@
+  }
+}
+
+function packBootstrapAddress(derivationPath, xpub, hdPassphrase, derivationScheme) {
```
Note: for testnet addresses this should also pack the protocol magic into the address

refi93 commented on features/address.js:
```diff
@@ -147,6 +147,7 @@ function unpackBootstrapAddress(address, hdPassphrase) {
   }
 
   return {
+    addressAttributes,
```
this should probably be `addressAttributes: attributes` (or rename `attributes` to `addressAttributes`)

refi93 replied:
fixed

refi93 commented:
I fixed unpackBootstrapAddress (it was failing because the return statement was referencing an undeclared variable) and updated packBootstrapAddress call to take also protocolMagic explicitly so it works properly even in testnet (testnet bootstrap addresses have the protocol magic explicitly encoded)

PeterBenc commented on features/address.js:
```diff
@@ -0,0 +1,252 @@
+  const addressAsBuffer = cbor.decode(base58.decode(address))[0].value
+  const addressData = cbor.decode(addressAsBuffer)
+  const addressAttributes = addressData[1]
+  const addressPayload = cbor.decode(addressAttributes.get(1))
```
@refi93  There seems to be a bug. First of all `addressAttributes.get(1)` throws an error, since `addressAttributes` isn't Map (at least not in case its empty)
If anything it should be something like this 
`const addressPayload = addressAttributes[1] ? cbor.decode(addressAttributes.get[1]) : something`. I am really not sure what to set as addressPayload in case its not in addressAtrributes.

refi93 replied:
good catch @PeterBenc - this function would work only for V1 (Daedalus addresses) which have non-empty attributes. Moreover, cbor lib seems to be decoding empty map to empty object which does not have the "get" attribute.

I gave it a second thought and the `unpackBootstrapAddress` call is imho misleading the way it currently works as its output is not the inverse of `packBootstrapAddress` which takes more parameters. So I decided to break the function down into `getBootstrapAddressAttributes`, `getBootstrapAddressDerivationPath` and `getBootstrapAddressProtocolMagic` and I refoctored the api to expect the addresses as buffers to stay consistent with similar calls for Shelley addresses. And I added the addressToBuffer utility function which translates string representation of addresses to buffers.

roccomuso commented:
Is there any issue running this on emsdk sdk-1.39.19?
I'm getting a lot of weird behaviors. (pheraps related to how wasm is loaded)
Does the lib needs to be initialized first now?

Like trying to do a simple:

```js
const sampleRootKeypairV1 = Buffer.from('d809b1b4b4c74734037f76aace501730a3fe2fca30b5102df99ad3f7c0103e48d54cde47e9041b31f3e6873d700d83f7a937bea746dadfa2c5b0a6a92502356ce6f04522f875c1563682ca876ddb04c2e2e3ae718e3ff9f11c03dd9f9dccf69869272d81c376382b8a87c21370a7ae9618df8da708d1a9490939ec54ebe43000', 'hex')
const sampleExtendedPublicKey = sampleRootKeypairV1.slice(64, 128)
lib.packBootstrapAddress([], sampleExtendedPublicKey, null, 2, MAINNET_PROTOCOL_MAGIC)
```

Will result first in a `Uncaught TypeError: Cannot read property 'malloc' of undefined` error. If you execute it again it will work but with a BAD buffer, thus generating a wrong address! executing it again will give again a bad result and the third time it returns the right buffer thus the right address when base58 encoding: `Ae2tdPwUPEZCxt4UV1Uj2AMMRvg5pYPypqZowVptz3GYpK4pkcvn3EjkuNH`.

refi93 commented on README.md:
```diff
@@ -38,41 +38,56 @@ var parentWalletPublicKey = parentWalletSecret.slice(64, 128)
+* `Map getBootstrapAddressAttributes(Buffer address)`
+* `Array<int>? getBootstrapAddressDerivationPath(Buffer address, Buffer hdPassphrase)`
+* `int getBootstrapAddressProtocolMagic(Buffer address)`
+* `Bool isValidBootstrapAddress(string address)`
```
it probably makes sense to expose a generic `isValidAddress` call which would use the correct validation logic depending on the address type (bootstrap/shelley)
